go/codegen/experimental/codegen_step.go

```diff
@@ -24,37 +26,121 @@ import (
-
-func generateStepCode(sqlStmt ir.SQLFlowStmt, stepIndex int, session *pb.Session) (string, error) {
+func generateStepCodeAndImage(sqlStmt ir.SQLFlowStmt, stepIndex int, session *pb.Session, sqlStmts []ir.SQLFlowStmt) (string, string, error) {
 	switch stmt := sqlStmt.(type) {
 	case *ir.TrainStmt:
-		return generateTrainCode(stmt, stepIndex, session)
+		return generateTrainCodeAndImage(stmt, stepIndex, session)
 	case *ir.PredictStmt:
-		return generatePredictCode(stmt, stepIndex, session)
+		return generatePredictCodeAndImage(stmt, stepIndex, session, sqlStmts)
 	case *ir.NormalStmt:
-		return GenerateNormalStmtStep(string(*stmt), session, stepIndex)
+		code, err := generateNormalStmtStep(string(*stmt), stepIndex, session)
+		return code, "", err
 	default:
-		return "", fmt.Errorf("not implemented stmt execution type %v", stmt)
+		return "", "", fmt.Errorf("not implemented stmt execution type %v", stmt)
+	}
+}
+
+func generateTrainCodeAndImage(trainStmt *ir.TrainStmt, stepIndex int, session *pb.Session) (string, string, error) {
+	image, isXGBoost, err := getImageAndIsXGBoostModel(trainStmt.Into, session, trainStmt)
+	if err != nil {
+		return "", "", err
+	}
+
+	if isXGBoost {
+		code, err := XGBoostGenerateTrain(trainStmt, stepIndex, session)
+		if err != nil {
+			return "", "", err
+		}
+		return code, image, nil
+	}
+	return "", "", fmt.Errorf("not implemented estimator type %s", trainStmt.Estimator)
+}
+
+func generatePredictCodeAndImage(predStmt *ir.PredictStmt, stepIndex int, session *pb.Session, sqlStmts []ir.SQLFlowStmt) (string, string, error) {
+	trainStmt := findModelGenerationTrainStmt(predStmt.Using, stepIndex, sqlStmts)
+	image, isXGBoost, err := getImageAndIsXGBoostModel(predStmt.Using, session, trainStmt)
+	if err != nil {
+		return "", "", err
+	}
+
+	if isXGBoost {
+		code, err := XGBoostGeneratePredict(predStmt, stepIndex, session)
+		if err != nil {
+			return "", "", err
+		}
+		return code, image, nil
+	}
+	return "", "", fmt.Errorf("not implemented model type")
+}
+
+// findModelGenerationTrainStmt finds the *ir.TrainStmt that generates the model named `modelName`.
+// TODO(sneaxiy): find a better way to do this when we have a well designed dependency analysis.
+func findModelGenerationTrainStmt(modelName string, idx int, sqlStmts []ir.SQLFlowStmt) *ir.TrainStmt {
+	idx--
+	for idx >= 0 {
+		trainStmt, ok := sqlStmts[idx].(*ir.TrainStmt)
+		if ok && trainStmt.Into == modelName {
+			return trainStmt
+		}
+		idx--
 	}
+	return nil
 }
 
-func generateTrainCode(trainStmt *ir.TrainStmt, stepIndex int, session *pb.Session) (string, error) {
-	if strings.HasPrefix(strings.ToUpper(trainStmt.Estimator), "XGBOOST.") {
-		return XGBoostGenerateTrain(trainStmt, stepIndex, session)
+func getImageAndIsXGBoostModel(modelName string, session *pb.Session, trainStmt *ir.TrainStmt) (string, bool, error) {
+	if trainStmt != nil {
+		return trainStmt.ModelImage, strings.HasPrefix(strings.ToUpper(trainStmt.Estimator), "XGBOOST."), nil
```

Comment:
Generating train code can write this line directly.

Reply:
Done.